### ISSUE TITLE
Memory-budget partitioner recomputes RNG ops with fresh randomness in backward, silently corrupting gradients at any activation_memory_budget < 1.0

### ISSUE THREAD

**Issue by HussainNizamani** [pytorch/pytorch#190758]
### 🐛 Describe the bug

When `torch.compile` is used with `torch._functorch.config.activation_memory_budget < 1.0`
and the model contains RNG ops (dropout, `randn_like`, ...), the partitioner can force
those ops to be recomputed in the backward graph, where they draw fresh randomness.
The backward then applies a different dropout mask than the forward actually used,
so gradients are silently wrong. No error, no warning.

This extends and corrects #190717, which reported the `budget == 0` case:

- `graphsafe_rng_functionalization` is not causally involved. It is only consulted
  inside `functionalize_rng_ops`, which never runs on this path: it is gated on
  user activation-checkpointing recompute tags (`partitioners.py:247-251, 4289-4293`),
  which partitioner-forced recompute never sets. The failure reproduces identically
  with the flag off, and on CPU.
- It is not only `budget == 0`. The `budget == 0` shortcut
  (`partitioners.py:3471-3472`) returns early and skips `solve_min_cut`'s
  "random op" recompute ban entirely — but at `0 < budget < 1` the knapsack path
  can *re-allow* recompute of the banned rng ops (`get_recomputable_banned_nodes`
  has no rng filter and `dont_ban` overrides the ban). I verified wrong gradients
  at budgets 0.0, 0.005 and 0.5.

### Minimal reproducer

Deterministic, and avoids conflating this bug with the documented eager-vs-inductor
Philox stream divergence — it checks the backward against the mask the compiled
forward itself applied:

```python
import torch
import torch.nn.functional as F
import torch._functorch.config as fc

fc.activation_memory_budget = 0.0  # also broken at 0.005 and 0.5

w = torch.arange(1.0, 26.0).view(5, 5)

def f(x):
    return F.dropout(x @ w, 0.5, True)

compiled = torch.compile(f)
torch.manual_seed(42)
x = torch.ones(2, 5, requires_grad=True)
out = compiled(x)
out.sum().backward()

mask = (out.detach() != 0).float()   # the mask the forward actually applied
expected = 2.0 * mask @ w.t()        # gradient implied by that mask
print(torch.equal(x.grad, expected)) # main: False — the backward redrew the mask
```

### Ablation

Triggers with the default inductor backend and with `backend="eager"`; with
`torch._inductor.config.fallback_random` True and False; at budgets 0.0, 0.005
and 0.5 (not at 1.0); on CPU builds and (per #190717's report) CUDA builds; also
with `torch.utils.checkpoint` (`use_reentrant=False`) combined with a low budget
(out-of-region rng ops), and with `randn_like` even in `eval()` mode.
`graphsafe_rng_functionalization` on/off makes no difference.

### Graph evidence

With `TORCH_LOGS=aot_graphs` at budget=0, the forward AND backward graphs each
draw their own independent seeds:

```
===== Forward graph 0 =====
    inductor_seeds_default: "i64[1][1]cpu" = torch.ops.prims.inductor_seeds.default(1, device(type='cpu'))
===== Backward graph 0 =====
    inductor_seeds_default: "i64[1][1]cpu" = torch.ops.prims.inductor_seeds.default(1, device(type='cpu'))
```

Full `TORCH_TRACE` log attached. With `fallback_random=True` the same happens
with `aten.native_dropout` recomputed directly in the backward.

### Training-level impact

Small two-dropout MLP, synthetic classification task, 5 seeds, CPU:

| configuration | mean eval acc | final train loss |
|---|---|---|
| eager | 0.894 | 0.224 |
| compiled, budget=0, with the fix below | 0.894 (identical curves with `fallback_random=True`) | 0.224 |
| compiled, budget=0, current main | 0.55–0.61 (near chance) | **NaN** |

### Fix

I have a two-part fix validated locally (CPU; the partitioner logic is
device-independent, CUDA validation via CI):

1. In the `budget == 0` shortcut, additionally save the tensor outputs of
   required-forward rng ops (op-level filter; all `getitem` projections for
   tuple-output ops). On the default path this saves only the small
   `inductor_seeds` tensor.
2. Exclude rng ops from the knapsack's tradeable candidates in
   `get_recomputable_banned_nodes`, so the solver's existing random-op ban stays
   final at every budget — as it already is at budget 1.0.

PR incoming with CPU regression tests (mask self-consistency at budgets 0 and
0.005, exact eager parity under `fallback_random=True`, and a backward-only-rng
case). Threading real RNG state through partitioner-forced recompute (extending
`functionalize_rng_ops` beyond AC tags) would be the fuller redesign — happy to
leave that call to maintainers.

The `budget == 0` shortcut predates graphsafe (#126318, May 2024).

[aot_graphs_buggy.log](https://github.com/user-attachments/files/30251903/aot_graphs_buggy.log)
[torch_trace_190717_buggy.log](https://github.com/user-attachments/files/30251902/torch_trace_190717_buggy.log)

### Error logs

No error is raised — that is the problem. Training runs normally and gradients
are silently wrong (backward applies a different dropout mask than the forward).
In our training test this collapsed accuracy to near chance with NaN loss.

### Versions

PyTorch version: 2.14.0a0+gitb761270
Is debug build: False
CUDA used to build PyTorch: None
ROCM used to build PyTorch: N/A
OS: Ubuntu 22.04.5 LTS (aarch64)
GCC version: (Ubuntu 11.4.0-1ubuntu1~22.04.3) 11.4.0
Clang version: Could not collect
CMake version: version 4.4.0
Libc version: glibc-2.35
Python version: 3.10.12 (main, Jun 22 2026, 18:55:27) [GCC 11.4.0] (64-bit runtime)
Python platform: Linux-6.8.0-1057-oracle-aarch64-with-glibc2.35
Is CUDA available: False
CUDA runtime version: No CUDA
GPU models and configuration: No CUDA
cuDNN version: No CUDA

cc @soulitzer @pbelevich @chauhang @penguinwu @bdhirsh @bobrenjc93 @aorenste
